src/skagent/model.py

```diff
@@ -13,7 +13,142 @@
+    1. Vector calibration: Arrays/lists of values for N agents
+    2. Distribution calibration: Distribution objects that get sampled for N agents
+
+    Parameters
+    ----------
+    calibration : dict
+        Dictionary of calibration parameters that may contain:
+        - Scalar values (float, int): kept as scalars unless heterogeneity is present
+        - Vectors (lists, arrays): used directly if length matches agent_count
+        - Distribution objects: sampled to generate agent-specific values
+
+    agent_count : int
+        Number of agents
+
+    rng : np.random.Generator, optional
+        Random number generator for sampling distributions
+
+    Returns
+    -------
+    dict
+        Calibration with heterogeneous parameters expanded to arrays when needed
+
+    Examples
+    --------
+    >>> from skagent.distributions import Normal
+    >>> # No heterogeneity - scalars preserved for backward compatibility
+    >>> calibration = {"DiscFac": 0.96, "CRRA": 2.0}
+    >>> expanded = expand_heterogeneous_calibration(calibration, agent_count=3)
+    >>> # Returns: {"DiscFac": 0.96, "CRRA": 2.0}
+    >>>
+    >>> # With heterogeneity - all parameters become arrays
+    >>> calibration = {"DiscFac": 0.96, "CRRA": [2.0, 2.5, 3.0]}
+    >>> expanded = expand_heterogeneous_calibration(calibration, agent_count=3)
+    >>> # Returns: {"DiscFac": [0.96, 0.96, 0.96], "CRRA": [2.0, 2.5, 3.0]}
+    Notes
+    -----
+    - Ex ante heterogeneity is applied once per agent and remains fixed over time
+      ("types").
+    - If no heterogeneity is specified anywhere in the calibration, scalar values
+      are preserved for backward compatibility.
+    - Discount factor heterogeneity (e.g., per-agent DiscFac) is excluded from the
+      scope of issue #91. This function is generic and will expand any parameter,
+      but project usage should keep discount factors homogeneous unless explicitly
+      supported elsewhere.
+
+    """
+    if rng is None:
+        rng = np.random.default_rng()
+
+    # Check if any parameter requires heterogeneity (is a vector or distribution)
+    has_heterogeneity = False
+    for param_value in calibration.values():
+        if isinstance(param_value, Distribution):
+            has_heterogeneity = True
+            break
+        elif isinstance(param_value, (list, tuple, np.ndarray)):
+            # Any list/array (even single element) indicates heterogeneity intention
+            has_heterogeneity = True
+            break
+
+    # If no heterogeneity is present, return original calibration for backward compatibility
+    if not has_heterogeneity:
+        return calibration.copy()
+
+    # Expand all parameters when heterogeneity is present
+    expanded = {}
+
+    for param_name, param_value in calibration.items():
+        if isinstance(param_value, Distribution):
+            # Distribution calibration: sample for each agent
+            expanded[param_name] = param_value.draw(agent_count)
```

Comment:
I think it might make sense to:
 - have an optional argument to`expand_heterogeneous_calibration` called `draw_args` or something better that's a dictionary, `{ sym: {arg_name : arg_val, ...}, ...}`
 - the values in this dictionary can be passed as keyword arguments to this draw function `param_value.draw(agent_count, **draw_args[param_name])` (needs a little more to handle when there are no provided `draw_args`
 - The use case here is if the user needs to pass in some special sampling requirements.

I wonder what you think of this? It's not critical and could be turned into a separate issue if it's not easy.